### ISSUE TITLE
WebSocketsProtocol doesn't use ManagedWebSocket optimized for ValueTask

### ISSUE THREAD

**Issue by buybackoff** [dotnet/corefx#29951]
ManagedWebSocket in [System.Net.WebSockets.WebSocketProtocol](https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs) is different from the implementation in [System.Net.WebSockets](https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs) for .NET 2.1. The former NuGet package is updated on May 5 as `4.5.0-rc1 (current)`, on the same day when [Microsoft.NETCore.App](https://www.nuget.org/packages/microsoft.netcore.app) package that includes the newest `System.Net.WebSockets` for .NET Core 2.1 was released as `2.1.0-rc1`. 

I had to dig into the source to find out that [Microsoft.AspNetCore.WebSockets](https://www.nuget.org/packages/Microsoft.AspNetCore.WebSockets), also released on that day as `2.1.0-rc1-final`, depends on `System.Net.WebSockets.WebSocketProtocol` and not on `System.Net.WebSockets` on [this line](https://github.com/aspnet/WebSockets/blob/ee4fec48e72daeb22f668789cfd08e94cf4948d3/src/Microsoft.AspNetCore.WebSockets/WebSocketMiddleware.cs#L122). Unfortunately, that doesn't allow to enjoy the performance benefits from https://github.com/dotnet/corefx/issues/27445 when running ASPNETCore 2.1 apps.

Is this for a reason? How to use the latest low-allocations implementation currently in ASPNETCore 2.1 apps?

**Comment by buybackoff** [comment]
I've just copied the package Microsoft.AspNetCore.WebSockets and added to the linked line:

```
#if NETCOREAPP2_1
                return WebSocket.CreateFromStream(opaqueTransport, isServer: true, subProtocol: subProtocol, keepAliveInterval: keepAliveInterval);
#else
                return WebSocketProtocol.CreateFromStream(opaqueTransport, isServer: true, subProtocol: subProtocol, keepAliveInterval: keepAliveInterval);
#endif
```
WebSocket now work as expected  in #27445 but the number of allocations is still very big.

![image](https://user-images.githubusercontent.com/489852/40648168-658d183a-6336-11e8-84d8-47fed44fd47f.png)

It looks like underlying socket allocates a lot and doesn't behave like in #27445.

Playing with my old [fork of SignalR](https://github.com/Spreads/SignalW) with the goal to eliminate amortized allocations "completely" on the server. Did SignalR achive that BTW? Cannot build the source master locally - too complicated without docs and with 2.2 dependencies and tooling. Ideally I want to understand if current SignalR adds enough overheads for small binary messages (I use my own header and dispatcher logic) so that manual WebSocket-only implementation could be better.

**Comment by davidfowl** [comment]
Yes, I think we need to cross somewhere here to get the correct overrides. Free performance win! Great catch!

/cc @BrennanConroy 

> Playing with my old fork of SignalR with the goal to eliminate amortized allocations "completely" on the server. Did SignalR achive that BTW? Cannot build the source master locally - too complicated without docs and with 2.2 dependencies and tooling. Ideally I want to understand if current SignalR adds enough overheads for small binary messages (I use my own header and dispatcher logic) so that manual WebSocket-only implementation could be better.

PS: Yes the current version of SignalR is much more efficient than the previous SignalW implementation. The version you forked was extremely early.

**Comment by buybackoff** [comment]
> PS: Yes the current version of SignalR is much more efficient than the previous SignalW implementation. The version you forked was extremely early.

For sure, but I cannot easily comprehend the current version and many abstractions between `bytes in  --> process them --> bytes out` 😄 

Looking into [HttpUpgradeStream](https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Kestrel.Core/Internal/Http/HttpUpgradeStream.cs) and do not see any methods with ValueTask. Probably optimized ManagedWebSocket is useless if the underlying stream can't help it?

**Comment by davidfowl** [comment]
I should say a bit more about SignalR though that's unrelated to this issue (which could be moved to https://github.com/aspnet/WebSockets/). 

> Playing with my old fork of SignalR with the goal to eliminate amortized allocations "completely" on the server.

Which allocations? Per connection or per message?

> Cannot build the source master locally - too complicated without docs and with 2.2 dependencies and tooling.

You can use the bits, they're on nuget and measure for yourself.

I'd continue the discussion about SignalR specifics over here https://github.com/aspnet/SignalR.

**Comment by davidfowl** [comment]
Sounds great! I'm pretty sure it'll be sufficient for your needs! The version you forked was a prototype. Lots of performance work went into the 1.0.0 release and there's more to go but most of the low hanging fruit was fixed.

> I need to compare it with some alternative. Maybe I end up with raw NetworkStream

That's pretty much writing your own web server which might not be crazy depending on what you're trying to achieve. If you're going down that path, you can look at this https://github.com/davidfowl/PlatformLevelTechempower/blob/944597f9ad196d3a333b837fb8c62a7304d5f9d2/ServerWithKestrel2.1/WebSocketConnection.cs. It's a websocket server built on top of kestrel directly without ASP.NET. It's still using the ManagedWebSocket implementation but it's up to you to decide how far you're willing to go.

> Sure when I touch & profile it. Issue for this repo is about ManagedWebSockets mismatch in sibling packages.

To put the issue back on topic, we either need to ifdef in ASP.NET or in the WebSocketProtocol package. I don't have a strong opinion here.

**Comment by davidfowl** [comment]
This is being fixed in ASP.NET Core 3.0. It's possible that a future version of `WebSocketProtocol.CreateFromStream` could be cross compiled to light up `WebSocket.CreateFromStream`. 

@stephentoub Any thoughts on that? If no, then we can close this bug?

**Comment by stephentoub** [comment]
> Any thoughts on that? If no, then we can close this bug?

I agree. WebSocketProtocol was created purely as a temporary solution to enable ASP.NET when built targeting netstandard20 to reuse the same WebSocket.CreateFromStream implementation without needing to copy and paste and tweak source.  When built targeting netstandard20, WebSocketProtocol couldn't use those new overloads introduced in .NET Core 2.1, and anything built for newer surface area can just target WebSocket.CreateFromStream.  So I agree this can be closed, as at this point I think WebSocketProtocol has served its purpose and is unlikely to evolve in any meaningful way.

**Comment by karelz** [comment]
Can we stop building it and shipping it then?

**Comment by karelz** [comment]
Triage: Stop building and shipping the assembly altogether.
